src/main/java/net/snowflake/client/core/bind/BindUploader.java

```diff
@@ -189,13 +196,126 @@ public static synchronized BindUploader newInstance(SFSession session, String st
+      int fileCount = 0;
+
+      while (rowNum < csvRows.size()) {
+        // create a list of byte arrays
+        while (numBytes < inputStreamBufferSize && rowNum < csvRows.size()) {
+          numBytes += csvRows.get(rowNum).length;
+          rowNum++;
+        }
+        // concatenate all byte arrays into 1 and put into input stream
+        ByteBuffer bb = ByteBuffer.allocate(numBytes);
+        for (int i = startIndex; i < rowNum; i++) {
+          bb.put(csvRows.get(i));
+        }
+        byte[] finalBytearray = bb.array();
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(finalBytearray);
+        // do the upload
+        String fileName = Integer.toString(++fileCount);
+        uploadStreamInternal(inputStream, fileName, true);
+        startIndex = rowNum;
+        numBytes = 0;
+        try {
+          inputStream.close();
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+  }
+
+  /**
+   * Method to put data from a stream at a stage location. The data will be uploaded as one file. No
+   * splitting is done in this method. Similar to uploadStreamInternal() in SnowflakeConnectionV1.
+   *
+   * <p>Stream size must match the total size of data in the input stream unless compressData
+   * parameter is set to true.
+   *
+   * <p>caller is responsible for passing the correct size for the data in the stream and releasing
+   * the inputStream after the method is called.
+   *
+   * @param inputStream input stream from which the data will be uploaded
+   * @param destFileName destination file name to use
+   * @param compressData whether compression is requested fore uploading data
+   * @throws SQLException raises if any error occurs
+   */
+  private void uploadStreamInternal(
+      InputStream inputStream, String destFileName, boolean compressData)
+      throws SQLException, BindException {
+
+    createStageIfNeeded();
+    String stageName = stagePath;
+    logger.debug("upload data from stream: stageName={}" + ", destFileName={}", destFileName);
+
+    if (stageName == null) {
+      throw new SnowflakeSQLLoggedException(
+          session,
+          ErrorCode.INTERNAL_ERROR.getMessageCode(),
+          SqlState.INTERNAL_ERROR,
+          "stage name is null");
+    }
+
+    if (destFileName == null) {
+      throw new SnowflakeSQLLoggedException(
+          session,
+          ErrorCode.INTERNAL_ERROR.getMessageCode(),
+          SqlState.INTERNAL_ERROR,
+          "stage name is null");
+    }
+
+    SFStatement stmt = new SFStatement(session);
+
+    StringBuilder putCommand = new StringBuilder();
+
+    // use a placeholder for source file
+    putCommand.append("put file:///tmp/placeholder ");
+
+    // add stage name
+    putCommand.append("'");
+    if (!(stageName.startsWith("@") || stageName.startsWith("'@") || stageName.startsWith("$$@"))) {
+      putCommand.append("@");
+    }
+    putCommand.append(stageName);
+    putCommand.append("'");
```

Comment:
Can we get reliable putCommand all the time? If stageName startsWith `'@`, the putCommand will be like `blabla ''@'`. Is this valid for backend to parse?

Reply:
I removed this whole piece of logic because the stage always starts with @. It is hardcoded earlier in the flow.